src/vision/status_quo/grace_wants_to_integrate_c_api.md

```diff
@@ -0,0 +1,176 @@
+similar to how the `Waker` object works in a `Future`'s `Context` type, in that it is how often the execution
+environment should re-try the future if the call to `poll` returns `Poll::Pending`.
+
+Something Grace learns to love immediately about Rust is that writing the `Future` here does not require her team to
+write their own execution environment. In fact, the future can be entirely written independent of the execution
+environment. She quickly writes an async method to represent the polling process:
+
+```rust
+/// Gets the next frame from the camera, waiting `retry_after` time until polling again if it fails.
+///
+/// Returns Some(frame) if a frame is found, or None if the camera is disconnected or goes down before a frame is
+/// available.
+async fn next_frame(camera: &Camera, retry_after: Duration) -> Option<Frame> {
+    while camera.is_available() {
+        if let Some(frame) = camera.poll() {
+            return Some(frame);
+        } else {
+            task::sleep_for(retry_after).await;
+        }
+    }
+
+    None
+}
+```
+
+The underlying C API doesn't provide any hooks that can be used to wake the `Waker` object on this future up, so Grace
+and her team decide that it is probably best if they just choose a sufficiently balanced `retry_after` period in which
+to try again. It does feel somewhat unsatisfying, as calling `sleep_for` feels about as hacky as calling
+`std::this_thread::sleep_for` in C++. However, there is no way to directly interoperate with the waker without having a
+separate thread of execution wake it up, and the underlying C library doesn't have any interface offering a notification
+for when that should be. In the end, this is the same kind of code that they would write in C, just without having to
+implement a custom execution loop themselves, so the team decides it is not a total loss.
+
+### The second problem: doing this many times
+
+Doing this a single time is fine, but an end goal of the project is to be able to stream frames from the camera for
+unspecified lengths of time. Grace spends some time searching, and realizes that what she actually wants is a `Stream`
+of some kind, so that her team can write something similar to:
+
+```rust
+let frame_stream = stream_from_camera(camera, Duration::from_millis(5));
+
+while let Some(frame) = frame_stream.next().await {
+    // process frames
+}
+
+println!("Frame stream closed.");
+```
+
+She scours existing crates, in particular looking for one way to transform the above future into a stream that can be
+executed many times. The only available option to transform a future into a series of futures is [`stream::unfold`],
+which seems to do exactly what Grace is looking for. Grace begins by adding a small intermediate type, and then plugging
+in the remaining holes:
+
+```rust
+struct StreamState {
+    camera: Camera,
+    retry_after: Duration,
+}
+
+
+
+fn stream_from_camera(camera: Camera, retry_after: Duration) -> Unfold<Frame, ??, ??> {
```

Comment:
Question (maybe something to add to the FAQ): Where did Grace get `Unfold` from? I guess she saw it in the Rustdoc, is that write?

Reply:
(I'm asking because I'm thinking a bit about what might have led her to write `impl Stream` from the get go)

Reply:
Firstly: Thanks for reading it, and thank you for the kind words! I appreciate the feedback.

Yes, I saw it in the type signature for `stream::unfold` in the futures crate. Initially I started writing everything as traits (which is perhaps another story) and using `impl Stream`as a return type in traits is... not currently viable. I eventually abandoned the trait architecture because async functions in traits are also problematic (and then get into the complexity of `Pin`, boxing your futures, etc.).

One thing that is somewhat tangential to this whole story is how traits are very much sub-optimal when working with async. Coming from C++ (and wanting some idea of virtual dispatch) it seems natural to use traits for everything, but that is just something to unlearn from C++, independent of async 🙃

Reply:
Can you maybe add a note about `Unfold` as a FAQ? Something lke

* Q: How did Grace know to use `Unfold` as the return type in the first place?
* She saw it in the rustdoc for `stream::unfold`.

Reply:
I added this to the FAQ, with a link to the rustdoc page in question!